File: Piloto/assets/js/compras.js
/* SGD — Módulo de Compras / Logística (cadastro de fornecedores).
 * Partilha a Ficha de Terceiro (terceiros.js). Ligado às contas correntes de fornecedores (32…). */
(function (global) {
  "use strict";
  const AY = global.AY;
  if (!AY) { console.error("compras.js: AY não carregado."); return; }
  const { read, write, uid, round2, hoje } = AY;

  const K = { fornecedores: AY.PREFIX + "cmp_fornecedores", compras: AY.PREFIX + "cmp_compras", docseq: AY.PREFIX + "cmp_docseq", seeded: AY.PREFIX + "cmp_seeded_v1" };

  function seed() {
    if (read(K.seeded, false)) return;
    write(K.fornecedores, [
      { id: uid("fo"), numero: "001", nome: "Distribuidora Central, Lda.", nif: "5410000001", localidade: "Luanda", telefone: "222 000 111", conta: "32121", tipoTerceiro: "Fornecedor", estado: "activo" },
      { id: uid("fo"), numero: "002", nome: "TecnoServe", nif: "5410000002", localidade: "Luanda", telefone: "222 000 222", conta: "32121", tipoTerceiro: "Fornecedor", estado: "activo" },
    ]);
    write(K.seeded, true);
  }
  function fornecedores() { seed(); return read(K.fornecedores, []); }
  function fornecedor(id) { return fornecedores().find(f => f.id === id) || null; }
  function proximoNumero() { let m = 0; fornecedores().forEach(f => { const n = parseInt(f.numero, 10); if (!isNaN(n)) m = Math.max(m, n); }); return String(m + 1).padStart(3, "0"); }
  function saveFornecedor(f) {
    const l = fornecedores();
    if (!f.id) { f.id = uid("fo"); f.criadoEm = hoje(); }
    if (!f.numero) f.numero = proximoNumero();
    const i = l.findIndex(x => x.id === f.id); if (i >= 0) l[i] = Object.assign(l[i], f); else l.push(f);
    write(K.fornecedores, l); return f;
  }
  function removeFornecedor(id) { write(K.fornecedores, fornecedores().filter(f => f.id !== id)); }

  // ---------- Documentos de Compra (fatura do fornecedor) ----------
  // O "tipo de documento" de uma compra É o documento contabilístico (o mesmo gerido em Diários e
  // Documentos) do Diário de Compras configurado — a lista mostra sempre todos os já criados, e
  // permite criar um novo no próprio acto do lançamento. Ao emitir, cada linha (obrigatoriamente
  // ligada a um artigo) gera uma entrada de stock via AY.log.registarMovimento, usando esse diário
  // e documento — essa única operação já debita as Existências, credita o Fornecedor (conta
  // corrente própria, criada/encontrada automaticamente pelo nome) e, se houver IVA, debita o IVA
  // Dedutível. Não há um lançamento de fatura em separado — a entrada em armazém É a contabilização
  // da compra, para não duplicar.
  function documentosCompra() {
    if (!AY.contab) return [];
    const diarioBase = (AY.log ? AY.log.cfg().diarioEntrada : null) || "21";
    return AY.contab.documentosDoDiario(diarioBase);
  }
  function comprasDocs() { seed(); return read(K.compras, []); }
  function compraDoc(id) { return comprasDocs().find(c => c.id === id) || null; }
  function nextNumeroCompra(cod) { cod = cod || "CP"; const seqs = read(K.docseq, {}); const n = (seqs[cod] || 0) + 1; seqs[cod] = n; write(K.docseq, seqs); return cod + " " + new Date().getFullYear() + "/" + String(n).padStart(4, "0"); }
  function calcTotaisCompra(linhas, ivaPerc) {
    const subtotal = round2((linhas || []).reduce((s, l) => s + (Number(l.qtd) || 0) * (Number(l.preco) || 0), 0));
    const iva = round2(subtotal * (Number(ivaPerc) || 0) / 100);
    return { subtotal, iva, total: round2(subtotal + iva) };
  }
  function saveCompraDoc(c) {
    if (!c.data) throw new Error("Indica a data do documento.");
    if (!c.documentoCod) throw new Error("Indica o tipo de documento.");
    const l = comprasDocs();
    c.linhas = (c.linhas || []).map(x => ({ artigoId: x.artigoId || "", unidade: x.unidade || "", descricao: x.descricao || "", qtd: Number(x.qtd) || 0, preco: round2(x.preco), total: round2((Number(x.qtd) || 0) * (Number(x.preco) || 0)) })).filter(x => x.descricao || x.total);
    const t = calcTotaisCompra(c.linhas, c.ivaPerc); c.subtotal = t.subtotal; c.iva = t.iva; c.total = t.total;
    if (!c.id) { c.id = uid("cp"); c.criadoEm = hoje(); c.estado = c.estado || "rascunho"; }
    const fo = fornecedor(c.fornecedorId); c.fornecedorNome = fo ? fo.nome : (c.fornecedorNome || "");
    const docRec = AY.contab ? AY.contab.documentos().find(d => d.codigo === c.documentoCod) : null;
    c.documentoNome = docRec ? docRec.descricao : (c.documentoNome || "");
    const i = l.findIndex(x => x.id === c.id); if (i >= 0) l[i] = Object.assign(l[i], c); else l.push(c);
    write(K.compras, l); return c;
  }
  function removeCompraDoc(id) { write(K.compras, comprasDocs().filter(c => c.id !== id)); }
  function compraEmitida(c) { return c.estado === "emitida"; }

  // Entrada de stock (por linha) ao emitir uma compra — cada linha tem obrigatoriamente um artigo,
  // já que o objectivo do documento é justamente movimentar o stock, além de lançar na contabilidade.
  function emitirCompra(id) {
    const c = compraDoc(id); if (!c) throw new Error("Documento inexistente.");
    if (compraEmitida(c)) throw new Error("Documento já emitido.");
    if (c.total <= 0) throw new Error("Documento sem valor.");
    if (!c.fornecedorId && !c.fornecedorNome) throw new Error("Indica o fornecedor.");
    if (!c.armazemId) throw new Error("Indica o armazém de entrada.");
    if (!c.documentoCod) throw new Error("Indica o tipo de documento.");
    const docRec = AY.contab.documentos().find(d => d.codigo === c.documentoCod);
    if (!docRec) throw new Error("O documento " + c.documentoCod + " já não existe — escolhe outro.");
    const linhasArt = (c.linhas || []).filter(l => l.qtd > 0);
    if (!linhasArt.length) throw new Error("Adiciona pelo menos uma linha.");
    if (linhasArt.some(l => !l.artigoId)) throw new Error("Todas as linhas têm de estar ligadas a um artigo — a compra tem de movimentar o stock.");
    if (!AY.log) throw new Error("Módulo de Logística indisponível.");
    if (!c.numero) c.numero = nextNumeroCompra(c.documentoCod);
    const movs = [], lancamentoIds = [], erros = [];
    linhasArt.forEach(l => {
      try {
        const mov = AY.log.registarMovimento({ tipo: "entrada", artigoId: l.artigoId, armazemId: c.armazemId, qtd: Number(l.qtd),
          custoUnit: l.preco, ivaPerc: c.ivaPerc || 0, data: c.data, documento: c.numero, entidade: c.fornecedorNome, descricao: "Compra " + c.numero,
          diarioContab: docRec.diario, documentoContab: docRec.codigo });
        movs.push(mov.id); if (mov.lancamentoId) lancamentoIds.push(mov.lancamentoId);
      } catch (e) { erros.push((l.descricao || "artigo") + ": " + e.message); }
    });
    if (!movs.length) throw new Error("Não foi possível movimentar o stock: " + erros.join(" · "));
    c.estado = "emitida"; c.stockMovs = movs; c.lancamentoIds = lancamentoIds; c.emitidoEm = hoje();
    saveCompraDoc(c);
    return { compra: c, stockMovs: movs, lancamentoIds, erros };
  }

  function resumoCompras() {
    const cs = comprasDocs(); const emit = cs.filter(compraEmitida);
    return { totalCompras: round2(cs.reduce((s, c) => s + c.total, 0)), totalRececionado: round2(emit.reduce((s, c) => s + c.total, 0)),
      nCompras: cs.length, nEmitidas: emit.length, porEmitir: round2(cs.filter(c => !compraEmitida(c)).reduce((s, c) => s + c.total, 0)) };
  }

  AY.compras = { K, fornecedores, fornecedor, saveFornecedor, removeFornecedor, proximoNumero,
    documentosCompra, comprasDocs, compraDoc, saveCompraDoc, removeCompraDoc, calcTotaisCompra, emitirCompra, compraEmitida, resumoCompras };
  try { seed(); } catch (e) { console.warn("compras seed:", e); }
})(window);

File: Producao/backend/src/services/compras.py
"""Compras: documentos de compra e entrada em armazém.

Transposto de `Piloto/assets/js/compras.js`.

Detalhe estrutural do Piloto que é preciso perceber: NÃO existe um lançamento de
factura de compra em separado. A entrada em armazém É a contabilização — uma só
operação que debita as Existências, credita a conta corrente do fornecedor e,
havendo IVA, debita o IVA Dedutível. Lançar a factura à parte duplicaria tudo.

É por isso que todas as linhas de uma compra têm de estar ligadas a um artigo.
"""

from decimal import ROUND_HALF_UP, Decimal
from uuid import UUID

from sqlalchemy import select
from sqlalchemy.orm import Session

from src.db.base import agora
from src.db.models.comercial import Compra
from src.db.models.contabilidade import DocumentoContabilistico
from src.services.contabilidade import ErroContabilistico
from src.services.comercial import proximo_numero
from src.services.notificacoes import notificar
from src.services.logistica import cfg_log, registar_movimento

ZERO = Decimal("0")
CENT = Decimal("0.01")


def r2(v) -> Decimal:
    return Decimal(v).quantize(CENT, rounding=ROUND_HALF_UP)


def documentos_compra(db: Session, empresa_id: UUID) -> list[DocumentoContabilistico]:
    """Tipos de documento disponíveis para uma compra.

    São os documentos contabilísticos do diário de entradas — a mesma tabela
    gerida em Diários e Documentos, não uma lista à parte.
    """
    diario = cfg_log(db, empresa_id).get("diario_entrada") or "21"
    return db.scalars(
        select(DocumentoContabilistico)
        .where(
            DocumentoContabilistico.empresa_id == empresa_id,
            DocumentoContabilistico.diario_codigo == diario,
            DocumentoContabilistico.ativo.is_(True),
        )
        .order_by(DocumentoContabilistico.codigo)
    ).all()


def calc_totais_compra(linhas: list[dict], iva_perc: Decimal) -> dict:
    subtotal = r2(
        sum(
            (Decimal(str(l.get("qtd") or 0)) * Decimal(str(l.get("preco") or 0))
             for l in linhas),
            ZERO,
        )
    )
    iva = r2(subtotal * (iva_perc or ZERO) / 100)
    return {"subtotal": subtotal, "iva": iva, "total": r2(subtotal + iva)}


def emitir_compra(
    db: Session, *, empresa_id: UUID, compra: Compra, exercicio_id: UUID | None = None
) -> dict:
    """Emite a compra: cada linha gera uma entrada de stock que contabiliza."""
    if compra.estado == "emitida":
        raise ErroContabilistico("Documento já emitido.")
    if (compra.total or ZERO) <= 0:
        raise ErroContabilistico("Documento sem valor.")
    if not compra.fornecedor_id and not compra.fornecedor_nome:
        raise ErroContabilistico("Indica o fornecedor.")
    if compra.armazem_id is None:
        raise ErroContabilistico("Indica o armazém de entrada.")
    if not compra.documento_codigo:
        raise ErroContabilistico("Indica o tipo de documento.")

    doc = db.scalar(
        select(DocumentoContabilistico).where(
            DocumentoContabilistico.empresa_id == empresa_id,
            DocumentoContabilistico.codigo == compra.documento_codigo,
        )
    )
    if doc is None:
        raise ErroContabilistico(
            f"O documento {compra.documento_codigo} já não existe — escolhe outro."
        )

    linhas = [l for l in compra.linhas if (l.qtd or ZERO) > 0]
    if not linhas:
        raise ErroContabilistico("Adiciona pelo menos uma linha.")
    if any(l.artigo_id is None for l in linhas):
        raise ErroContabilistico(
            "Todas as linhas têm de estar ligadas a um artigo — a compra tem de "
            "movimentar o stock."
        )

    if not compra.numero:
        compra.numero = proximo_numero(
            db, empresa_id, compra.documento_codigo, compra.data.year
        )

    movimentos, lancamentos, erros = [], [], []
    for l in linhas:
        try:
            m = registar_movimento(
                db, empresa_id=empresa_id, tipo="entrada", artigo_id=l.artigo_id,
                armazem_id=compra.armazem_id, qtd=l.qtd, custo_unit=l.preco,
                iva_perc=compra.iva_perc, data=compra.data, documento=compra.numero,
                entidade=compra.fornecedor_nome, descricao=f"Compra {compra.numero}",
                diario_contab=doc.diario_codigo, documento_contab=doc.codigo,
                exercicio_id=exercicio_id,
            )
            movimentos.append(m.id)
            if m.lancamento_id:
                lancamentos.append(m.lancamento_id)
        except ErroContabilistico as e:
            erros.append(f"{l.descricao or 'artigo'}: {e}")

    if not movimentos:
        raise ErroContabilistico(
            "Não foi possível movimentar o stock: " + " · ".join(erros)
        )

    compra.estado = "emitida"
    compra.documento_nome = doc.descricao
    compra.emitido_em = agora()

    # Notificação 3. Alguma linha entrou, por isso o documento é dado por
    # emitido — mas a mercadoria das linhas que falharam não está no
    # inventário nem nas contas, e o documento diz «emitida».
    if erros:
        notificar(
            db, empresa_id=empresa_id, capacidade="logistica.gerir",
            origem="compras", chave=f"compra-linhas-falhadas:{compra.id}",
            titulo=f"Compra {compra.numero}: {len(erros)} linha(s) não entraram em stock",
            texto=(
                "O documento foi emitido, mas estas linhas falharam: "
                + " · ".join(erros)
                + ". A mercadoria não está no inventário."
            ),
            ligacao="/logistica/compras",
            alvo_tipo="compra", alvo_id=compra.id,
        )

    db.flush()
    return {
        "compra_id": compra.id, "numero": compra.numero,
        "movimentos": movimentos, "lancamentos": lancamentos, "erros": erros,
    }


def resumo_compras(db: Session, *, empresa_id: UUID) -> dict:
    cs = db.scalars(select(Compra).where(Compra.empresa_id == empresa_id)).all()
    emit = [c for c in cs if c.estado == "emitida"]
    return {
        "total_compras": r2(sum((c.total or ZERO for c in cs), ZERO)),
        "total_rececionado": r2(sum((c.total or ZERO for c in emit), ZERO)),
        "n_compras": len(cs),
        "n_emitidas": len(emit),
        "por_emitir": r2(sum((c.total or ZERO for c in cs if c.estado != "emitida"), ZERO)),
    }
